From bd0117645f0c3d33d79fcb13a9a68a7c07bebbdf Mon Sep 17 00:00:00 2001
From: Guofeng Xu <guofxu@qti.qualcomm.com>
Date: Wed, 12 Aug 2026 14:28:41 +0800
Subject: [PATCH] journal: replace a bunch of assert() with friendlier checks

We should not rely that data stored in the journal files remains
entirely untouched at all times. Because we unallocate files, data might
go away any time. Hence, never assert() on any expectations on what the
file contains. Instead, handle it more gracefully as a corruption issue,
and return EBADMSG.

Signed-off-by: Lennart Poettering <lennart@poettering.net>
Git-commit: 5ee8b3edb385b216eb4f3316323ae1287824971a

Upstream-Status: Backport [https://github.com/systemd/systemd/commit/5ee8b3edb385b216eb4f3316323ae1287824971a]
---
 src/libsystemd/sd-journal/journal-file.c | 27 ++++++++++++++++++------
 1 file changed, 20 insertions(+), 7 deletions(-)

diff --git a/src/libsystemd/sd-journal/journal-file.c b/src/libsystemd/sd-journal/journal-file.c
index 08f3b8267b..00363e1545 100644
--- a/src/libsystemd/sd-journal/journal-file.c
+++ b/src/libsystemd/sd-journal/journal-file.c
@@ -2723,8 +2723,9 @@ static int bump_entry_array(
 
         if (direction == DIRECTION_DOWN) {
                 assert(o);
-                assert(o->object.type == OBJECT_ENTRY_ARRAY);
 
+                if (o->object.type != OBJECT_ENTRY_ARRAY)
+                        return -EBADMSG;
                 *ret = le64toh(o->entry_array.next_entry_array_offset);
         } else {
 
@@ -3234,9 +3235,11 @@ static int generic_array_bisect_for_data(
 
         assert(f);
         assert(d);
-        assert(d->object.type == OBJECT_DATA);
         assert(test_object);
 
+        if (d->object.type != OBJECT_DATA)
+                return -EBADMSG;
+
         n = le64toh(d->data.n_entries);
         if (n <= 0)
                 return 0;
@@ -3602,9 +3605,11 @@ int journal_file_move_to_entry_for_data(
 
         assert(f);
         assert(d);
-        assert(d->object.type == OBJECT_DATA);
         assert(IN_SET(direction, DIRECTION_DOWN, DIRECTION_UP));
 
+        if (d->object.type != OBJECT_DATA)
+                return -EBADMSG;
+
         /* FIXME: fix return value assignment. */
 
         /* This returns the first (when the direction is down, otherwise the last) entry linked to the
@@ -3664,7 +3669,9 @@ int journal_file_move_to_entry_by_offset_for_data(
 
         assert(f);
         assert(d);
-        assert(d->object.type == OBJECT_DATA);
+
+        if (d->object.type != OBJECT_DATA)
+                return -EBADMSG;
 
         return generic_array_bisect_for_data(
                         f,
@@ -3690,7 +3697,9 @@ int journal_file_move_to_entry_by_monotonic_for_data(
 
         assert(f);
         assert(d);
-        assert(d->object.type == OBJECT_DATA);
+
+        if (d->object.type != OBJECT_DATA)
+                return -EBADMSG;
 
         /* First, pin the given data object, before reading the _BOOT_ID= data object below. */
         r = journal_file_pin_object(f, d);
@@ -3756,7 +3765,9 @@ int journal_file_move_to_entry_by_seqnum_for_data(
 
         assert(f);
         assert(d);
-        assert(d->object.type == OBJECT_DATA);
+
+        if (d->object.type != OBJECT_DATA)
+                return -EBADMSG;
 
         return generic_array_bisect_for_data(
                         f,
@@ -3776,7 +3787,9 @@ int journal_file_move_to_entry_by_realtime_for_data(
 
         assert(f);
         assert(d);
-        assert(d->object.type == OBJECT_DATA);
+
+        if (d->object.type != OBJECT_DATA)
+                return -EBADMSG;
 
         return generic_array_bisect_for_data(
                         f,
-- 
2.43.0

